index.js

```diff
@@ -1,23 +1,42 @@
 /* jshint node: true */
 'use strict';
 
-var GithubAdapter = require('./lib/github');
-
-var nothing = function() {
-  return {
-    upload: function() {}
-  }
-}
+var Uploader = require('./lib/uploader');
+var DeployPluginBase = require('ember-cli-deploy-plugin');
 
 module.exports = {
-  name: 'ember-deploy-github',
-  type: 'ember-deploy-addon',
-  adapters: {
-    assets: {
-      'github': GithubAdapter
-    },
-    index: {
-      'github': nothing
-    }
+  name: 'ember-cli-deploy-github',
+  createDeployPlugin: function(options) {
+    var DeployPlugin = DeployPluginBase.extend({
+      name: options.name,
+      defaultConfig: {
+        branch: 'gh-pages',
+        gitClient: function(context) {
+          return require('nodegit');
+        },
+        distDir: function(context) {
+          return context.distDir;
+        }
+      },
+      requiredConfig: ['repository'],
+      configure: function(context) {
+        DeployPluginBase.prototype.configure.apply(this, arguments);
+
+        var options = {
+          branch:     this.readConfig('branch'),
+          repository: this.readConfig('repository')
+        };
+
+        var logger = this.log.bind(this);
+
+        var git = this.readConfig('gitClient');
+        this.uploader = new Uploader(git, options, logger);
```

Comment:
@dukex One way we have done this previously is to instantiate our objects like this (https://github.com/zapnito/ember-cli-deploy-revision-key/blob/master/index.js#L29-L31)

This gives the `Uploader` access to the plugin and it's `readConfig` and `log` functions directly. This will save pretty much all the logic you have in the `configure` hook.

Just a thought.